src/hotspot/cpu/x86/gc/g1/g1BarrierSetAssembler_x86.cpp

```diff
@@ -263,6 +276,40 @@ void G1BarrierSetAssembler::g1_write_barrier_pre(MacroAssembler* masm,
   __ bind(done);
 }
 
+static Assembler::Condition generate_single_region_test(MacroAssembler* masm, const Register store_addr, const Register new_val, const Register tmp) {
+  __ movptr(tmp, store_addr);                     // tmp := store address
+  __ xorptr(tmp, new_val);                        // tmp := store address ^ new value
+  __ shrptr(tmp, G1HeapRegion::LogOfHRGrainBytes);  // ((store address ^ new value) >> LogOfHRGrainBytes) == 0?
+  return Assembler::equal;
+}
+
+static Assembler::Condition generate_new_val_null_test(MacroAssembler* masm, const Register new_val) {
+  __ cmpptr(new_val, NULL_WORD);  // new value == null?
+  return Assembler::equal;
+}
+
+static Assembler::Condition generate_card_young_test(MacroAssembler* masm, const Register store_addr, const Register tmp, const Register tmp2) {
+  CardTableBarrierSet* ct = barrier_set_cast<CardTableBarrierSet>(BarrierSet::barrier_set());
+  __ movptr(tmp, store_addr);                                    // tmp := store address
+  __ shrptr(tmp, CardTable::card_shift());                       // tmp := card address relative to card table base
+  // Do not use ExternalAddress to load 'byte_map_base', since 'byte_map_base' is NOT
+  // a valid address and therefore is not properly handled by the relocation code.
+  __ movptr(tmp2, (intptr_t)ct->card_table()->byte_map_base());  // tmp2 := card table base address
```

Comment:
> Consider using lea instructions and ExternalAddress to generate relocation:

Just for my understanding, do you mean that the comment immediately above (introduced in [JDK-8028109](https://bugs.openjdk.org/browse/JDK-8028109)):
```
// Do not use ExternalAddress to load 'byte_map_base', since 'byte_map_base' is NOT
// a valid address and therefore is not properly handled by the relocation code.
```
 does not hold anymore in mainline?

> The same in generate_c1_post_barrier_runtime_stub()

Note that this JEP is not concerned with C1 barriers, in fact `generate_c1_post_barrier_runtime_stub()` is not touched by the changeset.

> Leyden needs relocation for card table base.

For ease of reviewing, porting, etc. I would suggest to introduce the required changes for project Leyden as a follow-up RFE, would that work?

Reply:
Okay.